### PR TITLE
Fix rounding error when casting timestamp to string for timestamps before 1970

### PR DESCRIPTION
**Pull request by andygrove**
This closes https://github.com/NVIDIA/spark-rapids/issues/868

### CODE REVIEW

**Comment by andygrove** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala [hunk 1]
```diff
@@ -491,10 +491,30 @@ case class GpuCast(
...
-      withResource(input.getBase.asStrings("%Y-%m-%d %H:%M:%S.%3f")) { cv =>
-        cv.stringReplaceWithBackrefs(GpuCast.TIMESTAMP_TRUNCATE_REGEX,"\\1\\2\\3")
-      })
+    // https://github.com/rapidsai/cudf/issues/5166
+    // The time is off by 1 second if the result is < 0
+    val adjustedTimestamp = withResource(input.getBase
+        .castTo(DType.TIMESTAMP_MICROSECONDS)) { micros =>
+      withResource(micros.castTo(DType.INT64)) { micros =>
+        withResource(Scalar.fromLong(1000000)) { oneSecond =>
+          withResource(micros.sub(oneSecond)) { plusOne =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala:500]: nit: I don't think this is `plusOne` when we subtract one second from it.

Comment (andygrove) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala]: oops .. thanks, fixed.

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

**Review by revans2 — APPROVED** [review]

File: tests/src/test/scala/com/nvidia/spark/rapids/CastOpSuite.scala [hunk 2]
```diff
@@ -692,7 +692,10 @@ object CastOpSuite {
     import session.sqlContext.implicits._
     val df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS")
     val timestampStrings = Seq(
-      "2020-12-31T11:59:59.999",
+      "1920-12-31T11:59:59.999",
```
Comment (razajafri) [inline tests/src/test/scala/com/nvidia/spark/rapids/CastOpSuite.scala:695]: can we add a few microseconds?

Comment (andygrove) [reply tests/src/test/scala/com/nvidia/spark/rapids/CastOpSuite.scala:694]: @razajafri Done. This actually highlighted a bug in the original implementation where we only supported timestamps with millisecond precision.

**Comment by andygrove** [comment]
build

**Review by razajafri — APPROVED** [review]

**Pull request merged**